crates/matrix-sdk/src/guaranteed_completion.rs

```diff
@@ -0,0 +1,254 @@
+// Copyright 2022 Famedly GmbH
+// Copyright 2022 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Optional methods guaranteeing a synchronized event store upon return.
+//!
+//! # How it works
+//! For each method with the suffix `_completed`, an event handler is
+//! registered before the sister function without the suffix is called.
+//! Upon return of the latter, a channel to the event handler receives
+//! events related to the action requested from the Matrix Server and
+//! the function returns only after the corresponding event has been
+//! registered.
+
+use ruma::{
+    api::client::membership::{join_room_by_id, join_room_by_id_or_alias},
+    events::{
+        room::member::{MembershipState, RoomMemberEventContent},
+        StateEvent, SyncStateEvent,
+    },
+    OwnedServerName, RoomId, RoomOrAliasId,
+};
+use tokio::sync::mpsc;
+use tracing::debug;
+
+use crate::{
+    error::HttpResult,
+    room::{Common, Room},
+    Client, Result,
+};
+
+impl Client {
+    /// Join a room by `RoomId` and return only after the corresponding
+    /// event has been synced. Requires a running Sync.
+    ///
+    /// Returns a `join_room_by_id::Response` consisting of the
+    /// joined rooms `RoomId`.
+    ///
+    /// # Arguments
+    ///
+    /// * `room_id` - The `RoomId` of the room to be joined.
+    pub async fn join_room_by_id_completed(
+        &self,
+        room_id: &RoomId,
+    ) -> HttpResult<join_room_by_id::v3::Response> {
+        let (tx, mut rx) = mpsc::unbounded_channel::<StateEvent<RoomMemberEventContent>>();
+
+        let handle = self.register_event_handler({
+            move |event: SyncStateEvent<RoomMemberEventContent>, room: Room| {
+
+                let tx = tx.clone();
+
+                async move {
+
+                    let full_event = event.into_full_event(room.room_id().to_owned());
+
+                    if full_event.membership() == &MembershipState::Join
+                        {
+                            if let Err(e) = tx.send(full_event){
```

Comment:
Also

```suggestion
                            if let Err(e) = tx.send(full_event) {
```

Seems like it's the "nested too deep for rustfmt" thingy. Maybe that could get moved out into helper methods. Same for some places below.